hx711 incompatible with stable rust

```
error[E0554]: `#![feature]` may not be used on the stable release channel
 --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/hx711-0.4.0/src/lib.rs:9:1
  |
9 | #![feature(never_type)]
  | ^^^^^^^^^^^^^^^^^^^^^^^
```

I'm not super sure what `never_type` does, is it possible to remove this constraint?

Oh thats what it does :(

```
error[E0658]: the `!` type is experimental
  --> hx711/src/lib.rs:61:11
   |
61 | impl Into<!> for Error<!, !> {
   |           ^
   |
   = note: see issue #35121 <https://github.com/rust-lang/rust/issues/35121> for more information
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,9 +1,12 @@
 language: rust
 
-matrix:
-  include:
-    - env: TARGET=x86_64-unknown-linux-gnu
-      rust: nightly
+env:
+  - TARGET=x86_64-unknown-linux-gnu
+  - TARGET=thumbv7m-none-eabi
+rust:
+  - nightly
+  - beta
+  - stable
 
 before_install:
   - set -e
@@ -21,6 +24,11 @@ cache: cargo
 before_cache:
   # Travis can't cache files that are not readable by "others"
   - chmod -R a+r $HOME/.cargo
+
+branches:
+  only:
+  - master
+  - testing
 
 notifications:
   email:

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,6 +8,9 @@ keywords = ["embedded-hal-driver", "loadcell", "hx711", "amplifier", "adc"]
 repository = "https://github.com/jonas-hagen/hx711"
 license = "MIT OR Apache-2.0"
 
+[features]
+never_type = []
+
 [dependencies]
 nb = "0.1.2"
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,5 +15,5 @@ never_type = []
 nb = "0.1.2"
 
 [dependencies.embedded-hal]
-version = "0.2.3"
+version = "0.2.4"
 features = ["unproven"]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "hx711"
-version = "0.4.0"
+version = "0.5.0"
 description = "A platform agnostic driver to interface with the HX711 (load cell amplifier and ADC)"
 authors = ["Jonas Hagen <jonas.hagen3@gmail.com>"]
 categories = ["embedded", "hardware-support", "no-std"]

--- a/README.md
+++ b/README.md
@@ -117,7 +117,7 @@ Some random notes on this topic:
 
 ## Changelog
 
-### v0.5 (next release)
+### v0.5
 
 - Hide usage of `never_type` behind feature for usage on stable rust
 

--- a/README.md
+++ b/README.md
@@ -20,6 +20,25 @@ Because the interface works by toggling of GPIO pins, some care has to be taken.
 
 See here: https://github.com/jonas-hagen/hx711-examples
 
+Incomplete appetizer:
+```rust
+let mut val: i32 = 0;
+
+let dout = gpioa.pa6.into_floating_input(&mut gpioa.crl);
+let pd_sck = gpioa.pa7.into_push_pull_output(&mut gpioa.crl);
+
+let mut hx711 = Hx711::new(Delay::new(cp.SYST, clocks), dout, pd_sck).into_ok();
+
+// Obtain the tara value
+writeln!(tx, "Obtaining tara ...").unwrap();
+const N: i32 = 8;
+for _ in 0..N {
+    val += block!(hx711.retrieve()).into_ok(); // or unwrap, see features below
+}
+let tara = val / N;
+writeln!(tx, "Tara:   {}", tara).unwrap();
+```
+
 ## Optional features
 
 ### `never_type`

--- a/README.md
+++ b/README.md
@@ -20,6 +20,26 @@ Because the interface works by toggling of GPIO pins, some care has to be taken.
 
 See here: https://github.com/jonas-hagen/hx711-examples
 
+## Optional features
+
+### `never_type`
+
+The `never_type` feature can optionally be enabled when using nightly rust.
+
+For some HALs, the digital input and output pins can never fail.
+When using the driver with such a crate, one can use `.into_ok()` on all results instead of `.unwrap()` or `.expect()`.
+
+Example with e.g. STM32f1xx embedded hal:
+```rust
+// Without never_type (stable rust):
+// We know that this never fails
+let weight = block!(hx711.retrieve()).unwrap())
+
+// With never_type (nightly rust)
+// It is obvious that this is always ok
+let weight = block!(hx711.retrieve()).into_ok()
+```
+
 ## Bit-banging and delays
 
 The protocol is implemented using the GPIO interface because the HX711 needs a specific number of clock cycles to set the operation mode (25, 26 or 27 cycles). 
@@ -77,6 +97,10 @@ Some random notes on this topic:
 * How to make this work on linux? Any ideas?
 
 ## Changelog
+
+### v0.5 (next release)
+
+- Hide usage of `never_type` behind feature for usage on stable rust
 
 ### v0.4
 

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -2,7 +2,13 @@ set -euxo pipefail
 
 main() {
     cargo check --target $TARGET
-    cargo test
+    case $TARGET in
+        x86_64*)
+            cargo test
+            ;;
+        *)
+            ;;
+    esac
 }
 
 main

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -1,10 +1,26 @@
 set -euxo pipefail
 
+run_tests() {
+    case $TRAVIS_RUST_VERSION in
+        stable)
+            cargo test
+            ;;
+        beta)
+            cargo test
+            ;;
+        nightly)
+            cargo test --features "never_type"
+            ;;
+        *)
+            ;;
+    esac
+}
+
 main() {
     cargo check --target $TARGET
     case $TARGET in
         x86_64*)
-            cargo test
+            run_tests
             ;;
         *)
             ;;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -112,7 +112,7 @@ where
     pub fn enable(&mut self) -> Result<(), Error<EIN, EOUT>> {
         self.pd_sck.set_low().map_err(Error::Output)?;
         self.delay.delay_us(TIME_SCK_LOW);
-        nb::block!{self.set_mode(self.mode)}
+        nb::block! {self.set_mode(self.mode)}
     }
 
     /// Reset the chip.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -6,7 +6,8 @@
 
 #![deny(missing_docs)]
 #![no_std]
-#![feature(never_type)]
+#![cfg_attr(feature = "never_type", feature(never_type))]
+#![cfg_attr(all(feature = "never_type", test), feature(unwrap_infallible))]
 
 extern crate embedded_hal as hal;
 
@@ -16,6 +17,7 @@ use hal::blocking::delay::DelayUs;
 use hal::digital::v2::InputPin;
 use hal::digital::v2::OutputPin;
 
+#[cfg(feature = "never_type")]
 use core::convert::Infallible;
 
 /// Maximum ADC value
@@ -59,6 +61,7 @@ pub enum Error<EIN, EOUT> {
 
 /// For some hardware crates, the digital input and output pins can never fail.
 /// This implementation enables the use of `.into_ok()`.
+#[cfg(feature = "never_type")]
 impl Into<!> for Error<!, !> {
     fn into(self) -> ! {
         panic!()
@@ -67,6 +70,7 @@ impl Into<!> for Error<!, !> {
 
 /// For some hardware crates, the digital input and output pins can never fail.
 /// This implementation enables the use of `.into_ok()`.
+#[cfg(feature = "never_type")]
 impl Into<!> for Error<Infallible, Infallible> {
     fn into(self) -> ! {
         panic!()
@@ -177,6 +181,9 @@ fn i24_to_i32(x: i32) -> i32 {
 mod tests {
     use crate::*;
 
+    #[cfg(feature = "never_type")]
+    use core::convert::Infallible;
+
     #[test]
     fn convert() {
         assert_eq!(i24_to_i32(0x000001), 1);
@@ -186,5 +193,19 @@ mod tests {
         assert_eq!(i24_to_i32(0xF00000), -1048576);
         assert_eq!(i24_to_i32(0x800000), -8388608);
         assert_eq!(i24_to_i32(0x7FFFFF), 8388607);
+    }
+
+    #[test]
+    #[cfg(feature = "never_type")]
+    fn infallible_into_ok() {
+        let this_is_ok: Result<usize, Error<Infallible, Infallible>> = Ok(77);
+        assert_eq!(this_is_ok.into_ok(), 77);
+    }
+
+    #[test]
+    #[cfg(feature = "never_type")]
+    fn never_fail_into_ok() {
+        let this_is_ok: Result<usize, Error<!, !>> = Ok(77);
+        assert_eq!(this_is_ok.into_ok(), 77);
     }
 }

